syntax = "proto3";
package protos.playbooks;

import "protos/base.proto";
import "google/protobuf/wrappers.proto";
import "google/protobuf/struct.proto";

// Task Execution Result Protos
message PlaybookMetricTaskExecutionResult {

  message Result {
    enum Type {
      UNKNOWN = 0;
      TIMESERIES = 1;
      TABLE_RESULT = 2;
    }

    message GroupByLabelValue {
      google.protobuf.StringValue name = 1;
      google.protobuf.StringValue value = 2;
    }

    message Timeseries {
      message LabeledMetricTimeseries {
        message Datapoint {
          sfixed64 timestamp = 1;
          google.protobuf.DoubleValue value = 2;
        }
        repeated GroupByLabelValue metric_label_values = 1;
        google.protobuf.StringValue unit = 2;
        repeated Datapoint datapoints = 3;
      }
      repeated LabeledMetricTimeseries labeled_metric_timeseries = 3;
    }

    message TableResult {
      message TableColumn {
        google.protobuf.StringValue name = 1;
        google.protobuf.StringValue type = 2;
        google.protobuf.StringValue value = 3;
      }
      message TableRow {
        repeated TableColumn columns = 1;
      }
      repeated TableRow rows = 1;
    }

    Type type = 1;
    oneof result {
      Timeseries timeseries = 2;
      TableResult table_result = 3;
    }
  }

  PlaybookMetricTaskDefinition.Source metric_source = 1;
  google.protobuf.StringValue metric_name = 2;
  google.protobuf.StringValue metric_expression = 3;
  Result result = 5;
}

message PlaybookDecisionTaskExecutionResult {

  message Result {
    enum Type {
      UNKNOWN = 0;
      ELSE_EVALUATION_CONDITION = 1;
      TIMESERIES_EVALUATION_CONDITION = 2;
    }

    message ElseEvaluation {
      google.protobuf.BoolValue evaluation = 1;
    }

    message TimeseriesEvaluation {
      google.protobuf.BoolValue evaluation = 1;
      TimeseriesEvaluationTask.Rule rule = 2;
      google.protobuf.DoubleValue value = 3;
    }

    Type type = 1;
    google.protobuf.StringValue next_task = 2;
    oneof result {
      ElseEvaluation else_evaluation_condition = 3;
      TimeseriesEvaluation timeseries_evaluation_condition = 4;
    }
  }

  google.protobuf.UInt64Value decision_task_id = 1;
  google.protobuf.StringValue decision_task_name = 2;
  Result result = 3;
}

message PlaybookDataFetchTaskExecutionResult {

  message Result {
    enum Type {
      UNKNOWN = 0;
      TABLE_RESULT = 1;
    }


    message TableResult {
      message TableColumn {
        google.protobuf.StringValue name = 1;
        google.protobuf.StringValue type = 2;
        google.protobuf.StringValue value = 3;
      }
      message TableRow {
        repeated TableColumn columns = 1;
      }
      google.protobuf.StringValue raw_query = 1;
      google.protobuf.UInt64Value total_count = 2;
      google.protobuf.UInt64Value limit = 3;
      google.protobuf.UInt64Value offset = 4;
      repeated TableRow rows = 5;
    }

    Type type = 1;
    oneof result {
      TableResult table_result = 2;
    }
  }

  google.protobuf.UInt64Value data_fetch_task_id = 1;
  google.protobuf.StringValue data_fetch_task_name = 2;
  PlaybookDataFetchTaskDefinition.Source data_source = 3;
  Result result = 4;
}

message PlaybookDocumentationTaskExecutionResult {

  message Result {
    enum Type {
      UNKNOWN = 0;
      MARKDOWN = 1;
    }

    message MarkdownResult {
      google.protobuf.Struct text = 1;
    }

    Type type = 1;
    oneof result {
      MarkdownResult markdown_result = 2;
    }
  }

  google.protobuf.UInt64Value documentation_task_id = 1;
  google.protobuf.StringValue documentation_task_name = 2;
  Result result = 3;
}

message PlaybookActionTaskExecutionResult {

  message Result {
    enum Type {
      UNKNOWN = 0;
      API_RESPONSE = 1;
    }

    message ApiResponse {
      google.protobuf.StringValue request_method = 1;
      google.protobuf.StringValue request_url = 2;
      google.protobuf.UInt64Value response_status = 3;
      google.protobuf.Struct response_headers = 4;
      google.protobuf.Struct response_body = 5;
      google.protobuf.Struct error = 6;
      google.protobuf.Struct metadata = 7;
    }

    Type type = 1;
    oneof result {
      ApiResponse api_response = 2;
    }
  }

  google.protobuf.UInt64Value action_task_id = 1;
  google.protobuf.StringValue action_task_name = 2;
  Result result = 3;
}

message PlaybookTaskExecutionResult  {
  google.protobuf.StringValue error = 1;
  oneof result {
    PlaybookMetricTaskExecutionResult metric_task_execution_result = 2;
    PlaybookDecisionTaskExecutionResult decision_task_execution_result = 3;
    PlaybookDataFetchTaskExecutionResult data_fetch_task_execution_result = 4;
    PlaybookDocumentationTaskExecutionResult documentation_task_execution_result = 5;
    PlaybookActionTaskExecutionResult action_task_execution_result = 6;
  }
}


// Metric Task Execution Protos
message PlaybookCloudwatchTask {
  enum TaskType {
    UNKNOWN = 0;
    METRIC_EXECUTION = 1;
    FILTER_LOG_EVENTS = 2;
  }

  message CloudwatchMetricExecutionTask {
    message Dimension {
      google.protobuf.StringValue name = 1;
      google.protobuf.StringValue value = 2;
    }
    google.protobuf.StringValue namespace = 1;
    google.protobuf.StringValue region = 2;
    google.protobuf.StringValue metric_name = 3;
    repeated Dimension dimensions = 4;
    google.protobuf.StringValue statistic = 5;
    google.protobuf.StringValue process_function = 6;
  }

  message CloudwatchFilterLogEventsTask {
    google.protobuf.StringValue region = 1;
    google.protobuf.StringValue log_group_name = 2;
    google.protobuf.StringValue filter_query = 3;
  }

  TaskType type = 1;
  oneof task {
    CloudwatchMetricExecutionTask metric_execution_task = 2;
    CloudwatchFilterLogEventsTask filter_log_events_task = 3;
  }
}

message PlaybookGrafanaTask {
  enum TaskType {
    UNKNOWN = 0;
    PROMQL_METRIC_EXECUTION = 1;
  }

  message PromQlMetricExecutionTask {
    message LabelValue {
      google.protobuf.StringValue name = 1;
      google.protobuf.StringValue value = 2;
    }
    google.protobuf.StringValue promql_expression = 1;
    repeated LabelValue promql_label_option_values = 2;
    google.protobuf.StringValue dashboard_uid = 3;
    google.protobuf.StringValue dashboard_title = 4;
    google.protobuf.StringValue panel_id = 5;
    google.protobuf.StringValue panel_title = 6;
    google.protobuf.StringValue process_function = 7;
    google.protobuf.StringValue panel_promql_expression = 8;
  }

  TaskType type = 1;
  google.protobuf.StringValue datasource_uid = 2;
  oneof task {
    PromQlMetricExecutionTask promql_metric_execution_task = 3;
  }
}

message PlaybookNewRelicTask {
  enum TaskType {
    UNKNOWN = 0;
    ENTITY_APPLICATION_GOLDEN_METRIC_EXECUTION = 1;
    ENTITY_DASHBOARD_WIDGET_NRQL_METRIC_EXECUTION = 2;
    NRQL_METRIC_EXECUTION = 3;
  }

  message EntityApplicationGoldenMetricExecutionTask {
    google.protobuf.StringValue application_entity_guid = 1;
    google.protobuf.StringValue application_entity_name = 2;
    google.protobuf.StringValue golden_metric_name = 3;
    google.protobuf.StringValue golden_metric_unit = 4;
    google.protobuf.StringValue golden_metric_nrql_expression = 5;
    google.protobuf.StringValue process_function = 6;
  }

  message EntityDashboardWidgetNRQLMetricExecutionTask {
    google.protobuf.StringValue dashboard_guid = 1;
    google.protobuf.StringValue dashboard_name = 2;
    google.protobuf.StringValue page_guid = 3;
    google.protobuf.StringValue page_name = 4;
    google.protobuf.StringValue widget_id = 5;
    google.protobuf.StringValue widget_title = 6;
    google.protobuf.StringValue widget_type = 7;
    google.protobuf.StringValue widget_nrql_expression = 8;
    google.protobuf.StringValue unit = 9;
    google.protobuf.StringValue process_function = 10;
  }

  message NRQLMetricExecutionTask {
    google.protobuf.StringValue metric_name = 1;
    google.protobuf.StringValue nrql_expression = 2;
    google.protobuf.StringValue unit = 3;
    google.protobuf.StringValue process_function = 4;
  }

  TaskType type = 1;
  oneof task {
    EntityApplicationGoldenMetricExecutionTask entity_application_golden_metric_execution_task = 2;
    EntityDashboardWidgetNRQLMetricExecutionTask entity_dashboard_widget_nrql_metric_execution_task = 3;
    NRQLMetricExecutionTask nrql_metric_execution_task = 4;
  }
}

message PlaybookDatadogTask {
  enum TaskType {
    UNKNOWN = 0;
    SERVICE_METRIC_EXECUTION = 1;
    QUERY_METRIC_EXECUTION = 2;
  }

  message ServiceMetricExecutionTask {
    google.protobuf.StringValue service_name = 1;
    google.protobuf.StringValue environment_name = 2;
    google.protobuf.StringValue metric_family = 3;
    google.protobuf.StringValue metric = 4;
    google.protobuf.StringValue process_function = 5;
  }

  message QueryMetricExecutionTask {
    repeated string queries = 1;
    google.protobuf.StringValue formula = 2;
    google.protobuf.StringValue process_function = 3;
  }

  TaskType type = 1;
  oneof task {
    ServiceMetricExecutionTask service_metric_execution_task = 2;
    QueryMetricExecutionTask query_metric_execution_task = 3;
  }
}

message PlaybookMetricTaskDefinition {
  enum Source {
    UNKNOWN = 0;
    CLOUDWATCH = 1;
    GRAFANA = 2;
    NEW_RELIC = 3;
    DATADOG = 4;
    GRAFANA_VPC = 5;
  }
  Source source = 1;
  oneof task {
    PlaybookCloudwatchTask cloudwatch_task = 6;
    PlaybookGrafanaTask grafana_task = 7;
    PlaybookNewRelicTask new_relic_task = 8;
    PlaybookDatadogTask datadog_task = 9;
  }
}


// Decision Task Execution Protos
enum EvaluationConditionFunction {
  UNKNOWN_ECF = 0;
  ECF_AVG = 1;
  ECF_SUM = 2;
  ECF_MIN = 3;
  ECF_MAX = 4;
}

enum EvaluationConditionOperator {
  UNKNOWN_ECO = 0;
  GREATER_THAN = 1;
  LESS_THAN = 2;
  GREATER_THAN_EQUAL = 3;
  LESS_THAN_EQUAL = 4;
  EQUAL = 5;
  NOT_EQUAL = 6;
}

message TimeseriesEvaluationTask {
  enum InputType {
    UNKNOWN = 0;
    METRIC_TIMESERIES = 1;
  }
  message Rule {
    enum Type {
      UNKNOWN_TEC = 0;
      ROLLING = 1;
      CUMULATIVE = 2;
    }
    Type type = 1;
    EvaluationConditionFunction function = 2;
    EvaluationConditionOperator operator = 3;
    google.protobuf.DoubleValue window = 4;
    google.protobuf.DoubleValue threshold = 5;
    google.protobuf.StringValue next_task = 6;
    repeated PlaybookMetricTaskExecutionResult.Result.GroupByLabelValue label_values = 7;
  }

  repeated Rule rules = 1;
  InputType input_type = 2;
  oneof input {
    PlaybookMetricTaskExecutionResult metric_timeseries_input = 3;
  }

}

message ElseEvaluationTask {
  google.protobuf.StringValue next_task = 1;
}

message PlaybookDecisionTaskDefinition {
  enum EvaluationType {
    UNKNOWN_ET = 0;
    ELSE = 1;
    TIMESERIES = 2;
  }

  EvaluationType evaluation_type = 1;
  oneof condition {
    ElseEvaluationTask else_evaluation_task = 2;
    TimeseriesEvaluationTask timeseries_evaluation_task = 3;
  }
}


// Documentation Task Execution Protos
message PlaybookDocumentationTaskDefinition {
  enum Type {
    UNKNOWN = 0;
    MARKDOWN = 1;
  }
  Type type = 1;
  google.protobuf.StringValue documentation = 2;
}


// Data Fetch Task Execution Protos
message PlaybookClickhouseDataFetchTask {
  google.protobuf.StringValue database = 1;
  google.protobuf.StringValue query = 2;
}

message PlaybookPostgresDataFetchTask {
  google.protobuf.StringValue database = 1;
  google.protobuf.StringValue query = 2;
}

message PlaybookEksDataFetchTask {
  enum CommandType {
    UNKNOWN = 0;
    GET_PODS = 1;
    GET_DEPLOYMENTS = 2;
    GET_EVENTS = 3;
    GET_SERVICES = 4;
  }
  CommandType command_type = 1;
  google.protobuf.StringValue description = 2;
  google.protobuf.StringValue region = 3;
  google.protobuf.StringValue cluster = 4;
  google.protobuf.StringValue namespace = 5;
}

message PlaybookSqlDatabaseConnectionDataFetchTask {
  google.protobuf.StringValue query = 1;
}

message PlaybookDataFetchTaskDefinition {
  enum Source {
    UNKNOWN = 0;
    CLICKHOUSE = 1;
    POSTGRES = 2;
    EKS = 3;
    SQL_DATABASE_CONNECTION = 4;
  }
  Source source = 1;
  google.protobuf.StringValue order_by_column = 2;
  google.protobuf.UInt64Value limit = 3;
  google.protobuf.UInt64Value offset = 4;
  oneof task {
    PlaybookClickhouseDataFetchTask clickhouse_data_fetch_task = 5;
    PlaybookPostgresDataFetchTask postgres_data_fetch_task = 6;
    PlaybookEksDataFetchTask eks_data_fetch_task = 7;
    PlaybookSqlDatabaseConnectionDataFetchTask sql_database_connection_data_fetch_task = 8;
  }
}


// Playbook Action Task Execution Protos
message PlaybookApiCallTask {
  enum Method {
    UNKNOWN = 0;
    GET = 1;
    POST = 2;
    PUT = 3;
    PATCH = 4;
    DELETE = 5;
  }
  Method method = 1;
  google.protobuf.StringValue url = 2;
  google.protobuf.Struct headers = 3;
  google.protobuf.Struct payload = 4;
  google.protobuf.UInt64Value timeout = 5;
  google.protobuf.Struct cookies = 6;
}

message PlaybookActionTaskDefinition {
  enum Source {
    UNKNOWN = 0;
    API = 1;
  }
  Source source = 1;
  google.protobuf.UInt64Value id = 2;
  google.protobuf.StringValue name = 3;
  google.protobuf.Struct metadata = 4;
  oneof task {
    PlaybookApiCallTask api_call_task = 5;
  }
}


// Playbook Protos
message PlaybookTaskDefinition {

  enum Type {
    UNKNOWN = 0;
    METRIC = 1;
    DECISION = 2;
    DATA_FETCH = 3;
    DOCUMENTATION = 4;
    ACTION = 5;
  }

  google.protobuf.UInt64Value id = 1;
  google.protobuf.StringValue name = 2;
  google.protobuf.StringValue description = 3;
  google.protobuf.StringValue notes = 4;
  google.protobuf.Struct global_variable_set = 5;
  Type type = 6;
  oneof task {
    PlaybookMetricTaskDefinition metric_task = 7;
    PlaybookDecisionTaskDefinition decision_task = 8;
    PlaybookDataFetchTaskDefinition data_fetch_task = 9;
    PlaybookDocumentationTaskDefinition documentation_task = 10;
    PlaybookActionTaskDefinition action_task = 11;
  }
}

message PlaybookStepDefinition {
  message ExternalLink {
    google.protobuf.StringValue name = 1;
    google.protobuf.StringValue url = 2;
  }
  google.protobuf.UInt64Value id = 1;
  google.protobuf.StringValue name = 2;
  google.protobuf.StringValue description = 3;
  google.protobuf.StringValue notes = 4;
  repeated ExternalLink external_links = 5;
  repeated PlaybookTaskDefinition tasks = 6;
}

message Playbook {
  google.protobuf.UInt64Value id = 1;
  google.protobuf.StringValue name = 2;
  google.protobuf.StringValue description = 3;
  google.protobuf.BoolValue is_active = 4;
  google.protobuf.StringValue created_by = 5;
  sfixed64 created_at = 6;
  sfixed64 last_run_at = 7;
  PlaybookExecutionStatusType status = 8;
  repeated PlaybookStepDefinition steps = 9;
  google.protobuf.Struct global_variable_set = 10;
  google.protobuf.BoolValue has_triggers = 11;
}


// Crud Protos
enum PlaybookExecutionStatusType {
  UNKNOWN_STATUS = 0;
  CREATED = 1;
  RUNNING = 2;
  FINISHED = 3;
  FAILED = 4;
}

message PlaybookExecutionLog {
  google.protobuf.UInt64Value id = 1;
  sfixed64 timestamp = 3;
  google.protobuf.StringValue playbook_run_id = 4;
  Playbook playbook = 5;
  PlaybookStepDefinition step = 6;
  PlaybookTaskDefinition task = 7;
  PlaybookTaskExecutionResult task_execution_result = 8;
}

message PlaybookExecution {
  google.protobuf.UInt64Value id = 1;
  google.protobuf.StringValue playbook_run_id = 2;
  Playbook playbook = 3;
  PlaybookExecutionStatusType status = 4;
  sfixed64 created_at = 5;
  sfixed64 started_at = 6;
  sfixed64 finished_at = 7;
  TimeRange time_range = 8;
  google.protobuf.StringValue created_by = 9;
  repeated PlaybookExecutionLog logs = 10;
}

message UpdatePlaybookOp {
  enum Op {
    UNKNOWN = 0;
    UPDATE_PLAYBOOK_NAME = 1;
    UPDATE_PLAYBOOK_STATUS = 2;
    UPDATE_PLAYBOOK = 3;
    UPDATE_PLAYBOOK_ALERT_OPS_TRIGGER_STATUS = 4;
  }

  message UpdatePlaybookName {
    google.protobuf.StringValue name = 1;
  }

  message UpdatePlaybookStatus {
    google.protobuf.BoolValue is_active = 1;
  }

  message UpdatePlaybook {
    Playbook playbook = 1;
  }

  message UpdatePlaybookAlertOpsTriggerStatus {
    google.protobuf.UInt64Value alert_ops_trigger_id = 1;
    google.protobuf.BoolValue is_active = 2;
  }

  Op op = 1;
  oneof update {
    UpdatePlaybookName update_playbook_name = 2;
    UpdatePlaybookStatus update_playbook_status = 3;
    UpdatePlaybook update_playbook = 4;
    UpdatePlaybookAlertOpsTriggerStatus update_playbook_alert_ops_trigger_status = 5;
  }
}